### PR TITLE
caddyhttp: Enable HTTP/3 by default

### PR DESCRIPTION
**Pull request by mholt**
This PR enables HTTP/3 by default (#3833), and implements the `RequireAddressValidation` callback as recommended in #3055.

Also removes experimental_http3 options/configuration except it does leave it in the JSON config for now, but marks it as deprecated.

Still need to add a way to disable HTTP/3, probably through the use of the existing protocols configuration parameter.

We will likely merge this sometime after the v2.5.0 release.

### CODE REVIEW

**Comment by mholt** [comment]
@marten-seemann I went ahead and implemented a very simple load statistic (number of currently active requests). Maybe it's not great, but it's so dang cheap and simple that I want to start with it and see how it goes. Let me know if I'm on the right track in terms of implementation, and making sure I have the logic right for `AcceptToken`. Thanks!

File: modules/caddyhttp/app.go [hunk 1]
```diff
@@ -346,30 +346,24 @@ func (app *App) Start() error {
 					tlsCfg := srv.TLSConnPolicies.TLSConfig(app.ctx)
 					ln = tls.NewListener(ln, tlsCfg)
 
-					/////////
-					// TODO: HTTP/3 support is experimental for now
-					if srv.ExperimentalHTTP3 {
```
Comment (francislavoie) [inline modules/caddyhttp/app.go:351]: Shouldn't we just add a new option `DisableHTTP3` and flip the boolean here?

Comment (mholt) [reply modules/caddyhttp/app.go:351]: Don't we already have facilities to configure protocols? It might be ALPN, and HTTP/3 supports the "h3" identifier in ALPN: https://quicwg.org/base-drafts/draft-ietf-quic-http.html#section-3.1

File: listeners.go [hunk 2]
```diff
@@ -87,12 +87,24 @@ func ListenPacket(network, addr string) (net.PacketConn, error) {
...
-func ListenQUIC(addr string, tlsConf *tls.Config) (quic.EarlyListener, error) {
+// a context other than context.Background is meaningless. If activeRequests
+// is not nil, it should represent the current load on the server.
+func ListenQUIC(addr string, tlsConf *tls.Config, activeRequests *int64) (quic.EarlyListener, error) {
 	lnKey := "quic/" + addr
 
 	sharedEl, _, err := listenerPool.LoadOrNew(lnKey, func() (Destructor, error) {
-		el, err := quic.ListenAddrEarly(addr, http3.ConfigureTLSConfig(tlsConf), &quic.Config{})
+		el, err := quic.ListenAddrEarly(addr, http3.ConfigureTLSConfig(tlsConf), &quic.Config{
+			AcceptToken: func(clientAddr net.Addr, token *quic.Token) bool {
```
Comment (marten-seemann) [inline listeners.go:97]: General explanation: When sending the first Initial packet to an unknown server, the client doesn't have any token, and thus won't send any. If the server decides to perform a Retry (which costs 1 RTT), it sends a Retry packet containing a token, which the client will then use on all subsequent Initial packets.
For reconnects to a server there's another case one needs to consider: During the connection, the server provides tokens to the client to use on future connections.
When receiving (and validating a token), a server probably wants to accept tokens provided in Retry packets for only a short time frame (10s or so), while tokens provided on a fully established connection would be valid for longer. The [quic.Token](https://pkg.go.dev/github.com/lucas-clemente/quic-go#Token) allows you to distinguish between these two cases.

You probably therefore want to:
1. First check if the server is under load. If it is not, not sending a token is probably ok. Better to save that 1 RTT! Return `true` in that case.
2. If the server is under load, it might make sense to require a token (maybe based on the `net.Addr`, if you want to go fancy). That means:
    * returning `false` if `token == nil`
    * else: validating the token. Have a look at the default token validation functions in quic-go for inspiration: https://github.com/lucas-clemente/quic-go/blob/332473668a992ce372c54c353b2bdc24344cf46e/server.go#L241-L259

**Review by marten-seemann — COMMENTED** [review]
Super excited to see this happening! Let me know if you need anything else, happy to help!

File: listeners.go (hunk 2 above)
Comment (mholt) [reply listeners.go:97]: @marten-seemann Thanks for the clarifications! That's really helpful.

Any chance we can/should reuse the default token validation function?

Comment (mholt) [reply listeners.go:97]: @marten-seemann I made a small update to this function based on your feedback. I think it'd be awesome if that `quic.defaultAcceptToken()` function was exported so that we could just call it from here and reuse that logic correctly.

Thanks for the guidance/help!

**Comment by mholt** [comment]
I've pushed a commit that polishes this feature a bit more:

- Deprecate the `server { protocol` sub-option in global options, including `allow_h2c` and `strict_sni_host`.
- Move `strict_sni_host` to sub-option of `server`
- Add `server { protocols` sub-option, which allows setting which HTTP versions to enable: `h1 h2 h2c h3` are allowed values.

This lets the user toggle precisely which HTTP versions they want to support, including HTTP/3-only, which is currently our oldest open issue: #1614. (Yay!)

Because the Go standard library does not let us disable HTTP/1.1, we would [probably have to implement our own server](https://stackoverflow.com/questions/51521141/how-to-completely-disable-http-1-x-support) to support _only_ HTTP/2 (or h2c) without HTTP/1.1 fallback. I actually tried it:

```go
// http2Server is a server that speaks only HTTP/2.
type http2Server struct {
	h1server *http.Server
	server   *http2.Server
	handler  http.Handler
	logger   *zap.Logger
}

func (h2 http2Server) Serve(ln net.Listener) error {
	h2.server = new(http2.Server)
	opts := &http2.ServeConnOpts{
		BaseConfig: h2.h1server,
		Handler:    h2.handler,
	}
	for {
		conn, err := ln.Accept()
		if err != nil {
			return err
		}
		go h2.serve(conn, opts)
	}
}

func (h2 http2Server) serve(conn net.Conn, opts *http2.ServeConnOpts) {
	defer func() {
		if r := recover(); r != nil {
			buf := make([]byte, 64<<10)
			buf = buf[:runtime.Stack(buf, false)]
			h2.logger.Error("panic",
				zap.String("remote", conn.RemoteAddr().String()),
				zap.Any("error", r),
				zap.String("stack", string(buf)))
		}
	}()
	defer conn.Close()
	h2.server.ServeConn(conn, opts)
}
```

but I could not get it to work. Curl complained, `curl: (1) Received HTTP/0.9 when not allowed` - which I don't understand, but did I spend much time looking into why. Anyway, this seems like a huge maintenance burden & undertaking, so I decided to not support HTTP/2-only servers for now. However, HTTP/1-only and HTTP/3-only configurations work fine!

This means HTTP/3 can be disabled, and we've made protocol configuration easy and consistent.

---

**In summary:** I believe have finished this feature, all except for the AcceptToken tweak mentioned above. (I would still like `quic.defaultAcceptToken()` either exported or always called even if we set our own. /cc @marten-seemann :smiley:) 

Once that is taken care of, I think we can merge this and ship Caddy with HTTP/3 enabled by default.

**Comment by TNQOYxNU** [comment]
> but I could not get it to work. Curl complained, curl: (1) Received HTTP/0.9 when not allowed - which I don't understand, but did I spend much time looking into why.

try request with `curl --http2-prior-knowledge`, curl will try to do [http upgrade](https://www.rfc-editor.org/rfc/rfc7540#section-3.2) from http1 first even with `--http2` option. h2c server's response has no http1 header, will be treated as http0.9.
I tested the code, it works on my machine.

**Comment by mholt** [comment]
@TNQOYxNU Ah, that's good to know! Thank you. I will give that a try next time I revisit this (soon).

File: caddyconfig/httpcaddyfile/serveroptions.go [hunk 3]
```diff
@@ -141,22 +140,51 @@ func unmarshalCaddyfileServerOptions(d *caddyfile.Dispenser) (any, error) {
 				}
 				serverOpts.ShouldLogCredentials = true
 
+			case "protocols":
+				protos := d.RemainingArgs()
+				for _, proto := range protos {
+					if proto != "h1" && proto != "h2" && proto != "h2c" && proto != "h3" {
```
Comment (francislavoie) [inline caddyconfig/httpcaddyfile/serveroptions.go:146]: Small thing, in other languages I tend to use an "array includes" pattern when I'm checking if a particular variable is one of many values. In Go, this hasn't been easy because slices don't have per-type functions to do this in a typesafe way, but with generics, we have a way now:

```go
import "golang.org/x/exp/slices"
```

```suggestion
					if slices.Contains([]string{"h1", "h2", "h2c", "h3"}, proto) {
```

If you don't want to pull this in just for this, that's fine, but I think this reads cleaner.

Comment (mholt) [reply caddyconfig/httpcaddyfile/serveroptions.go:146]: Oh, that's cool! I am not sure I'm ready for Caddy to use generics quite yet (I know they're one of Go's most thoroughly vetted features ever, I just want to wait a little longer), especially here where the gain is very minimal (and it adds a dependency for... string comparison) but I do see the value of this in general and I think when we are ready for generics, something like this would be a nice use case!

**Comment by mholt** [comment]
Linking this upstream issue as it is the last item remaining before we merge: https://github.com/lucas-clemente/quic-go/issues/3494 -- really appreciate how thoughtful Marten is being about it. :slightly_smiling_face:

**Comment by bt90** [comment]
https://github.com/lucas-clemente/quic-go/issues/2877 sounds like a possible bottleneck for servers with more traffic.

**Comment by mholt** [comment]
@bt90 Possibly! Although the upstream blockers appear to have been merged since then, so it's probably just a matter of time.

**Comment by mholt** [comment]
This branch is ready for final checks; I think we are ready to merge. Thank you @marten-seemann for your help and congrats on the progress!

**Comment by mholt** [comment]
@TNQOYxNU I will go ahead and merge this as a starting point. If we have requests for HTTP/2-only servers that have a legitimate need, we can revisit that feature. But for now I'll just see how it goes. Thanks again for verifying that!
